####-sparse-observable.md

```diff
@@ -0,0 +1,517 @@
+
+This can in theory be avoided by storing an explicit $2\times2$ matrix of coefficients for single-qubit term.
+For observables whose terms all involve only one or two qubits, the memory impacts are potentially acceptable, but the constant-factor scaling as the number of non-identity qubits increases is far poorer overall.
+Asymptotically, doing this would have a memory usage of 36 bytes per non-identity qubit + 8 bytes per term, compared to 5 bytes per non-identity qubit + 16 bytes per term, and would correspondingly be more expensive to compute with and to decompose to "measurement" operations.
+
+Supporting things like this efficiently in the observable is also perhaps misleading for the user; the _measurement terms_ needed to support things like $\lvert 0\rangle\langle 1\rvert$ still involve the exponential explosion (they go like $X + iY$), which could easily lead to a user getting needing / being charged for exponentially scaling runtime from an operator that appeared benign.
+
+
+## Questions
+
+### Should `SparseObservable` allow parametric observables?
+
+The definition given of `SparseObservables` limits the coefficients to be double-precision complex floats.
+`SparsePauliOp` technically allows observables to be parametrised in terms of Qiskit's symbolic `Parameter` class (or really any Python object).
+
+Should `SparseObservable` also support these uses?
+
+Pros of support:
+
+* More flexibility in what can be represented in the operators.
+
+Cons:
+
+* Operations on the type will be more complex, and the general case will need to hold the Python GIL in Rust space to act on the observables.
+* More complex code is needed from Rust space to handle mathematical manipulations efficiently in the happy path of "no Python-space parametrisation".
+
+Jake: personally I'd avoid this unless we have a really strong compelling use-case for giving it first-class support.
+A user can always work around this simply by splitting the terms of their sum into different broadcast axes in the estimator pub, then calculating the sums themselves, which gives them far more freedom.
+
+
+### What mathematical operations should `SparseObservable` support?
+
+From above, the minimum set is:
+
+* addition of two `SparseObservable`s
+* tensor product of two `SparseObservable`s
+* evolution of one `SparseObservable` by a single term
+* multiplication by complex scalars
+* structural equality of two `SparseObservable`s (structural not mathematical; it's highly inefficient to detect equality if the abstract terms form an overcomplete spanning set).
+
+There are several others that might be possible too.
+These operations could all fairly easily be supported:
+
+* Evolution of `SparseObservable` by another: this is completely doable, just naturally has quadratic complexity.
+* Tidy-up structural compaction of the operator: summing all terms that share the same abstract operator, removing zeros at some specified tolerance.
```

Comment:
I think this would be somewhat crucial, especially if we consider construction of operators to happen in an "iterative" manner by means of a sequence of mathematical operations.
The existing `SparsePauliOp` already has a part of its existing `.simplify` routine implemented in Rust, but performance still leaves room for improvement. I wonder how much this new operator completely implemented in Rust would benefit purely from the language advantage? It might be difficult to predict though :thinking: 

At the same time, could the CSR-like structure allow an alternative (read: more performant) approach for an implementation of `.simplify`?

Just thinking out loud here...

Reply:
For background, the high-level algorithm of `SparsePauliOp.simplify` is basically:
1. start with an empty hashmap of `pauli: coeff`
2. for each term of the sum, add the term into the hashmap, summing the coeffs if there's a match
3. create a new `SparsePauliOp` with each term in the hashmap, if the coeff is sufficiently far from 0

The _way_ we do that has a fair amount of fiddling around the edges that's making it less efficient than it could be, but asymptotically, the runtime complexity is already the best I can think of.

That said, the complexity still scales as $`\mathcal O(\text{qubits}\times\text{terms})`$.  The CSR-like form would be effectively the same, and so have the same asymptotic complexity. As a rule of thumb, though, it would be faster when there's less memory used (i.e. individually sparse terms), simply because there's no real mathematical structure / algorithmic trickery going on here, and the limit is mostly the iteration speed over all stored data.

For operator construction, one thing (mostly unrelated to this comment) to highlight: the way I've written `SparseObservable` here, it's growable in place, which means for some operations (`+` being a notable one), we can do it in-place and growing, which is very nice for iteration.  `evolve` is harder to do in-place, though.